### ISSUE TITLE
Route splitting without loosing type inference

### ISSUE THREAD

**Issue by artecoop** [elysiajs/elysia#138]
Hello all!
one of the most difficult situation, at least for me to switch to elysia from either hono or another node framework is the missing ability to split rounding code among files. I've seen some proposal, even file system routing. But for a simple situation (e.g.: 5/6 routes, each having some endpoints) is there is any workaround or guidance?

Thanks!

**Comment by Luisetepe** [comment]
Hi! Having probably the same problem here. It's very easy and clean yo create group with prefixes for every "resource". For example:
`
export const apiProducts = new Elysia({ prefix: '/products' })
    .use(apiSetup)
    .post('/', createProductHandler, createProductHandlerSchema)
    .get('/', getAllProductsHandler, getAllProductsHandlerSchema)
`

That would be my `index.ts` of the folder that is going to contain 1 file per route, with this file containing both the handler and the handler schema. The big problem right now is typing the handler function (tried several things with `Context`, `TypedRoute` and another couple of things, but no luck)

`
import { TypedRoute, t } from 'elysia'

export const createProductHandlerSchema = {
    body: t.Object({
        name: t.String({ error: 'name is required' }),
        pvp: t.Numeric({ error: 'pvp is required' }),
        categoryId: t.Numeric({ error: 'categoryId is required' })
    }),
    response: {
        200: t.Object({
            id: t.Numeric(),
            name: t.String(),
            pvp: t.Numeric(),
            categoryId: t.Numeric()
        }),
        400: t.String()
    }
}

export async function createProductHandler({ body, store }: ????) {
    // do something with 'body'
    return {
        id: 1,
        name: 'CocaCola',
        pvp: 10,
        categoryId: 4
    }
}
`

**Comment by Luisetepe** [comment]
Hi everyone again! Not ideal in my opinion (I'd love to have a single generic class for all my project) but I ended up surrendering to a per-handler typing, adding only what you need to the type. In the example above it would be only the body, and a type that I declared for my shared state store:

```typescript
import { ApiStore } from '@api/api.types'
import { Static } from '@sinclair/typebox'
import { Context, t } from 'elysia'

export const createProductHandlerSchema = {
    body: t.Object({
        name: t.String({ error: 'name is required' }),
        pvp: t.Numeric({ error: 'pvp is required' }),
        categoryId: t.Numeric({ error: 'categoryId is required' })
    }),
    response: {
        200: t.Object({
            id: t.Numeric(),
            name: t.String(),
            pvp: t.Numeric(),
            categoryId: t.Numeric()
        }),
        400: t.String()
    }
}

export async function createProductHandler({
    body,
    store
}: Context<
    { body: Static<typeof createProductHandlerSchema.body> },
    ApiStore
>) {
    // do something with 'body'
    return {
        id: 1,
        name: 'CocaCola',
        pvp: 10,
        categoryId: 4
    }
}
```

**Comment by Luisetepe** [comment]
Hi again, and sorry for the spam. I'm leaving the last comment for the record, but I've managed to come up with a simple type that "FOR NOW" is working for me. Full example:

`api.utils.ts`
```typescript
import { DbContext, dbPool } from '@infrastructure/persistence/db.utils'
import { Static, TSchema } from '@sinclair/typebox'
import Elysia, { Context } from 'elysia'

export type ApiStore = {
    dbPool: DbContext
}

export type ApiHandlerParams<Schema> = Context<
    {
        [Property in keyof Schema]: Schema[Property] extends TSchema
            ? Static<Schema[Property]>
            : Schema[Property]
    },
    ApiStore
>

export const apiSetup = new Elysia({ name: 'apiSetup' }).state('dbPool', dbPool)
```

`create-product.endpoint.ts`
```typescript
import { ApiHandlerParams } from '@api/api.utils'
import { t } from 'elysia'

export const createProductHandlerSchema = {
    body: t.Object({
        name: t.String({ error: 'name is required' }),
        pvp: t.Numeric({ error: 'pvp is required' }),
        categoryId: t.Numeric({ error: 'categoryId is required' })
    }),
    response: {
        200: t.Object({
            id: t.Numeric(),
            name: t.String(),
            pvp: t.Numeric(),
            categoryId: t.Numeric()
        }),
        400: t.String()
    }
}

export async function createProductHandler({
    body,
    store
}: ApiHandlerParams<typeof createProductHandlerSchema>) {
    // here 'body' and 'store' have proper intellisense and types
    return {
        id: 1,
        name: 'CocaCola',
        pvp: 10,
        categoryId: 4
    }
}
```

`index.ts`
```typescript
import { apiSetup } from '@api/api.utils'
import { Elysia } from 'elysia'
import {
    createProductHandler,
    createProductHandlerSchema
} from './create-product.endpoint'

export const apiProducts = new Elysia({ prefix: '/products' })
    .use(apiSetup)
    .post('/', createProductHandler, createProductHandlerSchema)
```

**Comment by sahibalejandro** [comment]
Having all endpoints in one file is a real deal breaker for many of us, even if you manage to split your routes into groups with multiple Elysia instances, there could be very large groups.

This is hurts development experience for sure, I hope some one smarter than me comes with a good fix, I've tried several ways to workaround this with no luck.

**Comment by gtramontina** [comment]
Reading the docs, the [dependency injection](https://elysiajs.com/patterns/dependency-injection.html) section _seems_ to attempt to address it, more specifically [here](https://github.com/elysiajs/documentation/blob/93116227015a28f13c7130e4c5505102e5a090fb/docs/patterns/dependency-injection.md#L57-L92).

I'm not entirely sure if I followed what's been suggested there, though. 🤔

**Comment by Luk4h** [comment]
Hello, guys!
Got myself in a similar problem, I was looking to type my body from the schema. After some looking at the types files and trying to find a solution... **I think I got something!** 🎉

Here is my sample code:
```ts
  private initializeRoutes() {
    const users = new Elysia({ prefix: '/auth' })
      .post('/register', this.register, RegisterBody)
  }

  private register = async ({ body, set }) => {
    const { emailAddress, passwordString, firstName, lastName, cpfNumber } = body;
    if ( await this.AuthService.checkIfUserExists(cpfNumber) )
      return new HttpException(400, 'Já existe um usuário com este CPF.');
    
    try {
      const newUser = await this.AuthService.create({ emailAddress, passwordString, firstName, lastName, cpfNumber });
      set.status = 201;
      return newUser;
    } catch (err: unknown) {
      if (err instanceof HttpException)
        return err;
      return new HttpException(500, `Ocorreu um erro interno: ${JSON.stringify(err)}}`);
    }
  }
```
Sadly as this is, all the parameters on the register function are *any*:
<img width="458" alt="image" src="https://github.com/elysiajs/elysia/assets/78048969/daef6365-6be9-4be9-8023-add204898928">

To type that you can simply put the parameter as **Context**:
<img width="793" alt="image" src="https://github.com/elysiajs/elysia/assets/78048969/b1830578-686a-4371-95f4-ab602b6d18ca">

But we enter in another problem as the body now is any:
<img width="484" alt="image" src="https://github.com/elysiajs/elysia/assets/78048969/c8cb859f-90b7-4586-89d7-49b26eae2aa1">

To solve that I found the **UnwrapRoute** Type, which can receive a typeof a schema and automatically types the body:
<img width="555" alt="image" src="https://github.com/elysiajs/elysia/assets/78048969/e8ddec0f-eb79-406c-8749-572b0fa983f5">

I haven't started working with this yet, as I just found this solution and got no type errors or anys.

### TLDR: `Context<UnwrapRoute<typeof Schema>>`

**Comment by Extarys** [comment]
I'm using `decorate` to inject prisma at the root level of my routes. I'm also looking for a way to pass the type to sub-routes/sub-files.

**Comment by thecodeassassin** [comment]
I assume this is also the reason why I'm getting this:

```
Argument of type '(context: Context) => Promise<string | null>' is not assignable to parameter of type 'Handler<MergeSchema<UnwrapRoute<InputSchema<never>, {}>, {}>, { request: {}; store: {}; }, "/user/metadata/:key">'.
  Types of parameters 'context' and 'context' are incompatible.
 ```
 
 when doing this:
 
 ```
 
app.get('/user/metadata/:key', metadata.get)
```

I really don't want to create a huge file or use any for context. Is there any fix for this?

**Comment by vastamaki** [comment]
This is a bit problematic, right now it's pretty much impossible to create a single file containing all route definitions and having the actual functions in separate files. Hopefully, this is something that will be fixed soon.

**Comment by asyncci** [comment]
That what I was looking for ! Bump ! 
Here is the same problem 
![image](https://github.com/elysiajs/elysia/assets/116139249/69efea84-44b7-45ed-9605-510712f8aa84)

**Comment by saurabhsri108** [comment]
Hi everyone, is there any solution for this situation?

**user.route.ts**
```ts
import { Elysia } from "elysia";
import { registerUser } from "../controllers/user.controller";
import { registerUserRouteGuard } from "../guards/user.guard";

export const userRoutes = new Elysia({ prefix: "/users" }).post(
	"/",
	registerUser,
	registerUserRouteGuard
);
```

**user.controller.ts**
```ts
import type { Context } from "elysia";

export const registerUser = ({ body }: Context) => {
	// get user details from the user
	const { username } = body;  // this does not works
	// validate user details
	// sanitize user details
	// check if user already exists
	// check if images, check for avatar
	// if images available, upload it to cloudinary
	// create user object - create entry in db
	// get the new user response without password and refresh token field
	// check if response is returned successfully - null check
	// return the response to the user
	return {};
};
```
**user.guard.ts**
```ts
import { t } from "elysia";

export const registerUserRouteGuard = {
	body: t.Object({
		username: t.String({ error: "Username is required" }),
		email: t.String({ error: "Email is required" }),
		fullname: t.String({ error: "Fullname is required" }),
		avatar: t.String({ error: "User profile picture is required" }),
		coverImage: t.String(),
		password: t.String({ error: "Password is required" }),
	}),
	response: {
		200: t.Object({}),
		400: t.String(),
	},
};
```

The problem is that the user.controller.ts file's body de-structuring is not typesafe. If I write it like below, it works:

**user.route.ts**
```ts
import { Elysia } from "elysia";
import { registerUser } from "../controllers/user.controller";
import { registerUserRouteGuard } from "../guards/user.guard";

export const userRoutes = new Elysia({ prefix: "/users" }).post(
	"/",
	({body}) => { 
         const {username} = body; // this now works
         return {}
        },
	registerUserRouteGuard
);
```

**Comment by saurabhsri108** [comment]
> Hi again, and sorry for the spam. I'm leaving the last comment for the record, but I've managed to come up with a simple type that "FOR NOW" is working for me. Full example:
> 
> `api.utils.ts`
> 
> ```ts
> import { DbContext, dbPool } from '@infrastructure/persistence/db.utils'
> import { Static, TSchema } from '@sinclair/typebox'
> import Elysia, { Context } from 'elysia'
> 
> export type ApiStore = {
>     dbPool: DbContext
> }
> 
> export type ApiHandlerParams<Schema> = Context<
>     {
>         [Property in keyof Schema]: Schema[Property] extends TSchema
>             ? Static<Schema[Property]>
>             : Schema[Property]
>     },
>     ApiStore
> >
> 
> export const apiSetup = new Elysia({ name: 'apiSetup' }).state('dbPool', dbPool)
> ```
> 
> `create-product.endpoint.ts`
> 
> ```ts
> import { ApiHandlerParams } from '@api/api.utils'
> import { t } from 'elysia'
> 
> export const createProductHandlerSchema = {
>     body: t.Object({
>         name: t.String({ error: 'name is required' }),
>         pvp: t.Numeric({ error: 'pvp is required' }),
>         categoryId: t.Numeric({ error: 'categoryId is required' })
>     })
> }
> 
> export async function createProductHandler({
>     body,
>     store
> }: ApiHandlerParams<typeof createProductHandlerSchema>) {
>     // here 'body' and 'store' have proper intellisense and types
>     return {
>         id: 1,
>         name: 'CocaCola',
>         pvp: 10,
>         categoryId: 4
>     }
> }
> ```
> 
> `index.ts`
> 
> ```ts
> import { apiSetup } from '@api/api.utils'
> import { Elysia } from 'elysia'
> import {
>     createProductHandler,
>     createProductHandlerSchema
> } from './create-product.endpoint'
> 
> export const apiProducts = new Elysia({ prefix: '/products' })
>     .use(apiSetup)
>     .post('/', createProductHandler, createProductHandlerSchema)
> ```

This solution worked for me FOR NOW, it seems.

**utils/apiSetup.ts**

```ts
import { Static, TSchema } from "@sinclair/typebox";
import Elysia, { Context } from "elysia";

export type ApiHandlerParams<Schema> = Context<{
	[Property in keyof Schema]: Schema[Property] extends TSchema
		? Static<Schema[Property]>
		: Schema[Property];
}>;

export const apiSetup = new Elysia({ name: "apiSetup" });
```

Then, I simply used it as the first middleware of the main Elysia app init:

```ts
import { cookie } from "@elysiajs/cookie";
import { cors } from "@elysiajs/cors";
import serverTiming from "@elysiajs/server-timing";
import staticPlugin from "@elysiajs/static";
import { swagger } from "@elysiajs/swagger";
import { Elysia } from "elysia";
import { compression } from "elysia-compression";
import { rateLimit } from "elysia-rate-limit";
import { apiSetup } from "./utils/apiSetup";

const app = new Elysia({ prefix: "/api/v1" })
	.use(apiSetup)
	.use(
		cors({
			origin: Bun.env.CORS_ORIGIN!,
			credentials: true,
		})
	)
	.use(
		rateLimit({
			responseMessage: "Max limit of 10 calls per minute exceeded",
			countFailedRequest: true,
		})
	)
	.use(cookie())
	.use(
		staticPlugin({
			assets: "public",
		})
	)
	.use(compression())
	.use(serverTiming())
	.use(swagger());

export default app;
```

**Comment by lnfel** [comment]
Please use the plugin pattern instead of TS gymnastics your way to type everything, all you would have in the end is a messy unmaintainable types.

https://elysiajs.com/essential/plugin.html

**Comment by SaltyAom** [comment]
For a future reference, please see https://elysiajs.com/patterns/mvc.html

**Comment by alpharder** [comment]
> For a future reference, please see https://elysiajs.com/patterns/mvc.html

With all respect, this isn't ergonomic at all

**Comment by Pedromigacz** [comment]
> For a future reference, please see https://elysiajs.com/patterns/mvc.html

Is it possible to just expose the types instead telling us to not do that? Trying to move to the recommended way of doing it raises so many other issues, for me at least.
I figured out how to make the typescript work for my use case. My only issue would be not being able to rely on Elysia types in the future.

**Comment by sekoyo** [comment]
You should use `new Elysia()...` and add the plugins you want type inference on *again*. Plugins should have a `name` (and optionally a `seed`), Elysia then knows not to instantiate them twice.

```ts
// myPlugin.ts
const myPlugin = new Elysia({ name: 'myplugin'})
    .decorate('greeting', 'Hello you')

// subRoutes.ts
const subRoutes = new Elysia()
    .use(myPlugin)
    .get('/greetme-again', ({ greeting }) => greeting)

// index.ts
const app = new Elysia()
    .use(myPlugin)
    .use(subRoutes)
    .get('/greetme', ({ greeting }) => greeting)
    .listen(3000)
```

**Comment by L-Mario564** [comment]
IDK if this is "the definitive solution" but thought I'd share how I personally handle splitting routes in a project of mine.

```ts
// base.ts
import { Elysia } from 'elysia';

// You can add whatever logic and state every route shares here
export const base = new Elysia().decorate('env', { ... }); // I can now access env vars via Elysia context

// routers/user.ts
export const userRouter = new Elysia({ prefix: '/user' })
  .use(base)
  .get('/', async (context) => {
    context.env; // This works
    // ... 
  });

// index.ts
import { Elysia } from 'elysia';
import { base } from './base';
import { userRouter } from './routers/user';

const app = new Elysia()
  .use(base)
  .use(userRouter)
  .listen(3001);

**Comment by SaltyAom** [comment]
Hi, just in case we have just added something that might help to the documentation:

https://elysiajs.com/essential/handler.html#typescript

**Comment by iansummerlin** [comment]
I have achieved something similar with the plugin pattern

`
// server.ts
new Elysia()
    .use(getContact);

// getContact.ts
export const getContact = new Elysia().get(
  "/contacts/:id",
  (context) => {
    const contactId = context.params.id; // ok
    const tuna = context.params.tuna // typescript error
    const body = context.body.john // typescript error
  },
  {
    params: t.Object({
      id: t.String(),
    })
  }
);
`

but I agree I would prefer this:

`
// server.ts
new Elysia()
      .get("/contacts/:id", getContact, getContactSchema);
`

I like seeing the routes

**Comment by ChoaibMouhrachAkera** [comment]
@SaltyAom 

Hey 

Im doing dependency injection this way

```ts
const app = createApp({
  db: createDBClient(databaseConnectionURL),
});
```

```ts

export const createApp = (config: { db: DBInstance }) => {
  const database = new Database(config.db);
  return new Elysia().decorate("database", database).use(router);
};

export type Context = InferContext<ReturnType<typeof createApp>>;
```

```ts
export const router = new Elysia({ prefix: "/api" })
  .use(authController)
 .use(courseController)
  .onError(({ error }) => {
    if (error instanceof Exception) {
      return { success: false, message: error.message };
    }

    console.error(error);

    return { success: false, message: SOMETHING_WENT_WRONG };
  });
```

```ts
export const courseController = new Elysia({ prefix: "/courses" })
  .get("/", async (context: Context) => {
    const courses = await context.database.course.all();

    return {
      data: courses.map((course) => course.data),
    };
  })
  .get("/:id", <T extends Context>(context: T) => {
    const course = context.database.course.findOrThrow({
      courseId: context.params.id
    });

    return {
      data: course,
    };
  })
```
I get 
`Property 'id' does not exist on type '{}'.` when hovering over `ts       courseId: context.params.id`

**Comment by SaltyAom** [comment]
Update

It's recommended to tied type to the Elysia instance, however in the new version we do have [InferContext](https://elysiajs.com/essential/handler.html#infercontext) which should have solved this problem

Otherwise, please refers to [Best practice: service](https://elysiajs.com/essential/best-practice.html#service)

**Comment by VityaSchel** [comment]
> Update
> 
> It's recommended to tied type to the Elysia instance, however in the new version we do have [InferContext](https://elysiajs.com/essential/handler.html#infercontext) which should have solved this problem
> 
> Otherwise, please refers to [Best practice: service](https://elysiajs.com/essential/best-practice.html#service)

@SaltyAom Looks like `InferContext` was deprecated. Any chance you can reopen this issue? I'd like to split all routes to separate files but keep TypeScript hints. `.use()` doesn't work for me because TypeScript enters endless loop and eventually shows `'ElysiaInstance' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.`

```ts
// src/index.ts
const app = new Elysia()
  .get("/", () => "Welcome!") // GET /
  .use(usersRouter)
  .listen(3000)

// src/users/index.ts
const usersRouter = new Elysia({ prefix: "/users" }) // /users
  .get("/", () => "(List of users)") // GET /users
  .post("/broadcast", () => "(Broadcast to all users)") // POST /users/broadcast
  .use(userRouter)

// src/users/[id]/index.ts
const userRouter = new Elysia({ prefix: "/:id" }) // /users/:id
  .get("/", ({ params: { id } }) => `A particular user ${id}`) // GET /users/:id
  .patch("/", ({ params: { id } }) => `Modify user ${id}`) // PATCH /users/:id
  .delete("/", ({ params: { id } }) => `Modify user ${id}`) // DELETE /users/:id
  .use(userFoobarsRouter)
  .use(userBarbazesRouter)

// src/users/[id]/foobars/index.ts
const userFoobarsRouter = new Elysia({ prefix: "/foobars" }) // /users/:id/foobars
  .get("/", ({ params: { id } }) => `TypeScript doesn't see ${id} anymore!`) // GET /users/:id/foobars
  .post("/", ({ params: { id } }) => `TypeScript doesn't see ${id} anymore!`) // POST /users:/:id/foobars

// src/users/[id]/barbazes/index.ts
const userBarbazesRouter = new Elysia({ prefix: "/barbazes" }) // /users/:id/barbazes
  .get("/", ({ params: { id } }) => `TypeScript doesn't see ${id} anymore!`) // GET /users/:id/barbazes
  .post("/", ({ params: { id } }) => `TypeScript doesn't see ${id} anymore!`) // POST /users:/:id/barbazes
// ... etc ...
```

In this example, if you make a GET request to localhost:3000/users/1234/foobars — you'll get `TypeScript doesn't see 1234 anymore!` which is expected. But in development, TypeScript does not inherit `params` and all other properties usually inferred automatically.

Both `InferContext<typeof userRouter>` and `.use(userRouter)` on `userBarbazesRouter` result in "[X] is referenced directly or indirectly in its own type annotation." error.

I understand this is a simplified example, but you can imagine how these imaginary simple routes turn into hundreds of lines of code each which suggests using separate modules/typescript files for each route for quick navigation in the project regardless of its size.

**Comment by mehrdad-shokri** [comment]
I'm dealing with same problem. InferContext isn't documented. Any suggestions on this?
